src/ocean/math/WideInt.d

```diff
@@ -0,0 +1,474 @@
+
+        Enables assignment from a plain integer type
+
+        Params:
+            rhs = value to assign
+
+    ***************************************************************************/
+
+    public void opAssign ( ulong rhs )
+    {
+        this.payload[] = 0;
+        this.payload[0] = cast(uint) (rhs & uint.max);
+        this.payload[1] = cast(uint) (rhs >> 32);
+    }
+
+    unittest
+    {
+        WideInt i;
+        i = 42;
+        test!("==")(i.payload[0], 42);
+        i = ulong.max;
+        test!("==")(i.payload[0], uint.max);
+        test!("==")(i.payload[1], uint.max);
+    }
+
+    unittest
+    {
+        WideInt a = 42;
+        WideInt b = 43;
+        a = b;
+        test!("==")(a, 43);
+    }
+
+    /***************************************************************************
+
+        Enables equality comparison with a plain integer type
+
+        Params:
+            rhs = value to compare to
+
+    ***************************************************************************/
+
+    public equals_t opEquals ( ulong rhs )
+    {
+        if (this.payload[0] != (rhs & uint.max))
+            return false;
+
+        if (this.payload[1] != (rhs >> 32))
+            return false;
+
+        foreach (ref word; this.payload[2 .. N])
+        {
+            if (word != 0)
+                return false;
+        }
+
+        return true;
+    }
+
+    unittest
+    {
+        WideInt i;
+
+        i = 42;
+        test!("==")(i, 42);
+
+        i = ulong.max / 2;
+        test!("==")(i, ulong.max / 2);
+        test!("!=")(i, ulong.max);
+
+        i.payload[$-1] = 42;
+        test!("!=")(i, 42);
+    }
+
+    /***************************************************************************
+
+        Enables equality comparison with WideInt of the same size
+
+        Params:
+            rhs = value to compare against
+
+    ***************************************************************************/
+
+    public equals_t opEquals ( WideInt rhs )
+    {
+        return this.payload[] == rhs.payload[];
+    }
+
+    unittest
+    {
+        WideInt a, b;
+        test!("==")(a, b);
+        foreach (size_t i, ref elem; a.payload)
+        {
+            a.payload[i] = 1337;
+            b.payload[i] = 1337;
+        }
+        test!("==")(a, b);
+        a.payload[$-1] = 0;
+        test!("!=")(a, b);
+    }
+
+    /***************************************************************************
+
+        Increment current value by one fitting in uint
+
+        Params:
+            rhs = value to increment with. Limited to `uint` for now to simplify
+                internal arithmetic.
+
+        Throws:
+            WideIntRangeException if this WideInt value range was to be
+            exceeded
+
+    ***************************************************************************/
+
+    public void opAddAssign ( uint rhs )
+    {
+        if (add(this.payload[0], rhs))
```

Comment:
Oh, I see.  I think I just didn't look properly at what `checkAndInc` was doing before, so didn't follow either the consequences or the reason for the `enforce`.  Thanks again for clarifying.